File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/08-Stack/LeetCode/LeetCode-0020.py
class Solution:
    def isValid(self, s: str) -> bool:
        S = []
        for ch in s:
            if ch == '(' or ch == '[' or ch == '{':
                S.append(ch)
            else:
                if len(S) == 0: return False
                top = S.pop()

                if ch == ')' and top != '(': return False
                if ch == ']' and top != '[': return False
                if ch == '}' and top != '{': return False
        return len(S) == 0

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/08-Stack/LeetCode/LeetCode-0020.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> S = new Stack<>();

        for (int i=0; i<s.length(); i++){
            char ch = s.charAt(i);
            if (ch == '(' || ch == '[' || ch == '{')
                S.push(ch);
            else {
                if (S.isEmpty()) return false;
                char top = S.pop();

                if (ch == ')' && top != '(') return false;
                if (ch == ']' && top != '[') return false;
                if (ch == '}' && top != '{') return false;
            } //end-else
        } //end-for

        return S.empty();
    }
}